File: swift/347-Top-k-frequent-elements.swift
class Solution {
    func topKFrequent(_ nums: [Int], _ k: Int) -> [Int] {
        var count = [Int:Int]()
        for n in nums {
            guard count[n] != nil else {
                count[n] = 1
                continue
            }
            count[n] = 1 + count[n]!
        }
        
        var freq = [[Int]](repeating: [], count: nums.count+1)
        for (n, c) in count {
            freq[c].append(n)
        }
        
        var res = [Int]()
        for i in stride(from: freq.count-1, to: 0, by: -1) {
            for n in freq[i] {
                res.append(n)
                if res.count == k {
                    return res
                }
            }
        }
        return res
    }
}

File: cpp/347-Top-K-Frequent-Elements.cpp
class Solution {
public:
	vector<int> topKFrequent(vector<int>& nums, int k) {
		unordered_map<int, int> count;
		for (int n : nums) {
			++count[n];
		}

		vector<vector<int>> freq(nums.size() + 1, vector<int>());
		for (auto e : count) {
			freq[e.second].push_back(e.first);
		}

		vector<int> result;
		for (int i = freq.size()-1; i >= 0; --i) {
			for (int n : freq[i]) {
				result.push_back(n);
				if (result.size() == k)
					return result;
			}
		}
		return result;
	}
};
